### PR TITLE
[Relay][AlterOp] NHWC to NCHWc support for Pool, concatenate, sum.

### PR DESCRIPTION
**Pull request by anijain2305**
For quantized InceptionV3, the number of layout transforms are reduced with this PR

Original - 260 Layout transforms / 1836 total operators
After PR - 172 Layout transforms / 1718 total operators
(Many of the transforms out of 172 are cheaper than the original 260 LT).

### CODE REVIEW

**Comment by anijain2305** [comment]
@yzhliu @vinx13 @ZihengJiang @tqchen Please review when you get time.

File: src/relay/op/nn/pad.cc [hunk 1]
```diff
@@ -36,6 +36,78 @@ namespace relay {
...
+      axis_pad_width.emplace(axis->var->name_hint, params->pad_width[index_counter]);
+      index_counter++;
+    }
+
+    // 2) Create new pad width by walking over the new layout and using the map.
+    tvm::Array<tvm::Array<tvm::Expr>> new_pad_width;
+    for (auto axis : new_in_layouts[0]->axes) {
+      auto axis_name = axis->var->name_hint;
+      if (axis_pad_width.count(axis_name) != 0) {
+        // This is primal axis. So, directly use the original pad_width.
```
Comment (yzhliu) [inline src/relay/op/nn/pad.cc]: there's no guarantee that the old layouts are all primal, right?

File: src/relay/op/tensor/reduce.cc [hunk 2]
```diff
@@ -119,6 +119,56 @@ Array<Integer> GetExcludeAxes(size_t indim,
   return r_axes;
 }
 
+Array<Array<Layout>> SumInferCorrectLayout(const Attrs& attrs, const Array<Layout>& new_in_layouts,
```
Comment (yzhliu) [inline src/relay/op/tensor/reduce.cc]:
```suggestion
Array<Array<Layout>> ReduceInferCorrectLayout(const Attrs& attrs, const Array<Layout>& new_in_layouts,
```

File: src/relay/op/tensor/reduce.cc [hunk 3]
```diff
@@ -119,6 +119,56 @@ Array<Integer> GetExcludeAxes(size_t indim,
...
+    std::string new_layout_string = "";
+    int axis_index = 0;
+    for (auto layout_axis : new_in_layouts[0]->axes) {
+      char layout_dim = layout_axis->var->name_hint.at(0);
+      if (old_r_dims.count(layout_dim)) {
+        new_r_axis.push_back(tvm::Integer(axis_index));
+      }
+      // Collect only the primal axis.
+      if (LayoutAxis::Get(layout_axis).IsPrimal()) {
+        new_layout_string += layout_dim;
```
Comment (yzhliu) [inline src/relay/op/tensor/reduce.cc]: how about the sub-axis?

File: src/relay/op/tensor/reduce.cc [hunk 4]
```diff
@@ -119,6 +119,56 @@ Array<Integer> GetExcludeAxes(size_t indim,
...
+    for (auto r_axis : r_axes) {
+      old_r_dims.emplace(old_in_layouts[0].name().at(r_axis));
+    }
+
+    // 2) Collect the new axes by walking new_layout.
+    tvm::Array<tvm::Integer> new_r_axis;
+    std::string new_layout_string = "";
+    int axis_index = 0;
+    for (auto layout_axis : new_in_layouts[0]->axes) {
+      char layout_dim = layout_axis->var->name_hint.at(0);
```
Comment (yzhliu) [inline src/relay/op/tensor/reduce.cc]: would be better to consistently use LayoutAxis

File: tests/python/relay/test_pass_alter_op_layout.py [hunk 5]
```diff
@@ -580,6 +629,278 @@ def expected():
     assert(analysis.alpha_equal(a, b))
 
 
+def test_alter_layout_pad():
+    """ Check NCHW, NHWC and corner case for pad layout conversion"""
+    @register_alter_op_layout("nn.conv2d", level=112)
```
Comment (tmoreau89) [inline tests/python/relay/test_pass_alter_op_layout.py]: can we add a comment to explain the "level" value?

File: tests/python/relay/test_pass_alter_op_layout.py [hunk 6]
```diff
@@ -370,9 +370,18 @@ def expected():
 
     assert analysis.alpha_equal(a, b), "Actual = \n" + str(a)
 
+
 def test_alter_layout_concatenate():
-    """ """
-    def before():
+    """ NCHW, NHWC and corner case concatenate layout transform."""
+    @register_alter_op_layout("nn.conv2d", level=107)
```
Comment (tmoreau89) [inline tests/python/relay/test_pass_alter_op_layout.py]: can we add a comment to explain the "level" value?

File: tests/python/relay/test_pass_alter_op_layout.py [hunk 7]
```diff
@@ -580,6 +629,278 @@ def expected():
...
+
+    b = expected()
+    b = run_opt_pass(b, transform.InferType())
+
+    assert analysis.alpha_equal(a, b), "Actual = \n" + str(a)
+
+
+def test_alter_layout_pool():
+    """ Check NCHW, NHWC pool layout conversion"""
+    @register_alter_op_layout("nn.conv2d", level=113)
```
Comment (tmoreau89) [inline tests/python/relay/test_pass_alter_op_layout.py]: can we add a comment to explain the "level" value?

File: src/relay/op/tensor/reduce.cc (hunk 3 above)
Comment (anijain2305) [reply src/relay/op/tensor/reduce.cc]: Sub axis, we dont have to do anything. Basically, this will insert a layout transform from NCHWC16c to NCHW before reduce.

File: src/relay/op/tensor/reduce.cc (hunk 2 above)
Comment (anijain2305) [reply src/relay/op/tensor/reduce.cc]: Done

File: src/relay/op/tensor/reduce.cc (hunk 4 above)
Comment (anijain2305) [reply src/relay/op/tensor/reduce.cc]: Done

File: src/relay/op/nn/pad.cc [hunk 8]
```diff
@@ -36,6 +36,80 @@ namespace relay {
...
+  bool is_layout_modified = new_in_layouts.defined();
+  if (new_in_layouts.defined()) {
+    // Create a map of axis to param_width. For the new layout, a new param_width is generated using
+    // the map. The new layout is rejected, if the padding is happening along the axis which was
+    // split.
+
+    // 1) Create a map from axis to param_width using old layout.
+    std::map<std::string, tvm::Array<tvm::Expr>> axis_pad_width;
+    int index_counter = 0;
+    CHECK_EQ(new_in_layouts.size(), 1);
```
Comment (yzhliu) [inline src/relay/op/nn/pad.cc]: `CHECK_EQ(old_in_layouts.size(), 1);`

File: src/relay/op/nn/pad.cc [hunk 9]
```diff
@@ -36,6 +36,80 @@ namespace relay {
...
+      const auto& new_layout_axis = LayoutAxis::Get(iter_var);
+      auto axis_name = new_layout_axis.name();
+      if (axis_pad_width.count(axis_name) != 0 && new_layout_axis.IsPrimal()) {
+        // This is primal axis. So, directly use the original pad_width.
+        new_pad_width.push_back(axis_pad_width.at(axis_name));
+      } else {
+        // This is the axis that got split. So, check that pad_width was [0, 0] originally.
+        const auto& dual_axis = new_layout_axis.ToPrimal();
+        auto dual_axis_name = dual_axis.name();
+        CHECK(axis_pad_width.count(dual_axis_name));
```
Comment (yzhliu) [inline src/relay/op/nn/pad.cc]: suggest to print `old_in_layouts[0]` and `new_layout_axis` for easier debug.

File: src/relay/op/nn/pad.cc [hunk 10]
```diff
@@ -36,6 +36,80 @@ namespace relay {
...
+        new_pad_width.push_back(axis_pad_width.at(dual_axis_name));
+
+        // If all pad_width elements are not zero, do not change the layout.
```
Comment (yzhliu) [inline src/relay/op/nn/pad.cc]:
```suggestion
        // If any pad_width element is not zero, do not change the layout.
```

**Comment by yzhliu** [comment]
@anijain2305 Could you take a look at the failure test?

**Comment by anijain2305** [comment]
I am getting some coreml flaky errors with pad operator. I will try to reproduce them. Currently, removing Pad changes from this PR. Will send separately.

**Comment by anijain2305** [comment]
@yzhliu @zhiics can we get this merged? I will send a PR for pad separately.

**Comment by zhiics** [comment]
I will TAL later.

File: src/relay/op/tensor/reduce.cc [hunk 11]
```diff
@@ -119,6 +119,59 @@ Array<Integer> GetExcludeAxes(size_t indim,
   return r_axes;
 }
 
+// Return the modified layout for AlterOpLayout pass.
+Array<Array<Layout>> ReduceInferCorrectLayout(const Attrs& attrs,
+                                              const Array<Layout>& new_in_layouts,
+                                              const Array<Layout>& old_in_layouts,
+                                              const Array<Array<IndexExpr>>& old_in_shapes) {
+  // NOTE: Discard "const" qualifier here.
+  ReduceAttrs* params = const_cast<ReduceAttrs*>(attrs.as<ReduceAttrs>());
```
Comment (zhiics) [inline src/relay/op/tensor/reduce.cc]: Why const_cast here? is it because we need to change the axis?

File: python/tvm/relay/frontend/tflite.py [hunk 12]
```diff
@@ -748,10 +748,12 @@ def convert_conv(self, op, conv_type):
         elif padding == Padding.SAME:
             pad_top, pad_bottom = get_pad_value(input_h, dilated_kernel_h, stride_h)
             pad_left, pad_right = get_pad_value(input_w, dilated_kernel_w, stride_w)
-            in_expr = _op.nn.pad(data=in_expr, pad_width=((0, 0),
-                                                          (pad_top, pad_bottom),
-                                                          (pad_left, pad_right),
-                                                          (0, 0)))
+            do_pad = not (pad_top == 0 and pad_bottom == 0 and pad_left == 0 and pad_right == 0)
```
[the remark this answers is not in the document]
Comment (anijain2305) [reply python/tvm/relay/frontend/tflite.py]: This one is fine. It is just an optimization that ensures that we do not redundant pads.

File: src/relay/op/tensor/reduce.cc (hunk 11 above)
Comment (anijain2305) [reply src/relay/op/tensor/reduce.cc]: Yes.

**Comment by zhiics** [comment]
Thanks @anijain2305 @yzhliu @tmoreau89. This is now merged.
